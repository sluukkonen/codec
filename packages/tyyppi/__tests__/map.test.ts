import { fromJson, map, number, string } from "../src/index.js"
import { expectParseFailure, expectParseSuccess } from "./helpers.js"

test("should parse a map with valid key and value types", () => {
  expectParseSuccess(map(string, number), new Map())
  expectParseSuccess(map(string, number), new Map([["a", 1]]))
})

test("should reject values that are not maps", () => {
  expectParseFailure(map(string, number), [])
  expectParseFailure(map(string, number), null)
})

test("should reject a map with the wrong key type", () => {
  expectParseFailure(map(string, number), new Map([[1, 1]]))
})

test("should reject a map with the wrong value type", () => {
  expectParseFailure(map(string, number), new Map([["a", "a"]]))
})

test("should work if the key codec not simple", () => {
  expectParseSuccess(
    map(fromJson.bigint, number),
    new Map([["1", 1]]),
    new Map([[1n, 1]]),
  )
})

test("should work if the value codec not simple", () => {
  expectParseSuccess(
    map(number, fromJson.bigint),
    new Map([[1, "1"]]),
    new Map([[1, 1n]]),
  )
})
