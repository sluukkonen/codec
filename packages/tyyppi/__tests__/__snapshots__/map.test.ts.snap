// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should reject a map with the wrong key type 1`] = `
[
  {
    "code": "invalid_type",
    "expected": "string",
    "message": "Expected a string, received a number",
    "path": [],
    "received": "number",
  },
]
`;

exports[`should reject a map with the wrong value type 1`] = `
[
  {
    "code": "invalid_type",
    "expected": "number",
    "message": "Expected a number, received a string",
    "path": [],
    "received": "string",
  },
]
`;

exports[`should reject values that are not maps 1`] = `
[
  {
    "code": "invalid_type",
    "expected": "map",
    "message": "Expected a map, received an array",
    "path": [],
    "received": "array",
  },
]
`;

exports[`should reject values that are not maps 2`] = `
[
  {
    "code": "invalid_type",
    "expected": "map",
    "message": "Expected a map, received null",
    "path": [],
    "received": "null",
  },
]
`;
